### PR TITLE
Add some tests for Xapi_vm_lifecycle.check_operation_error

### PR DESCRIPTION
**Pull request by gaborigloi**
And update the operations put into the `VM.allowed_operations` field by `VM.update_allowed_operations` - there were many VM operations missing that the function didn't even consider.

### CODE REVIEW

File: ocaml/xapi/test_vm_check_operation_error.ml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+
+let with_test_vm f =
+  let __context = Mock.make_context_with_new_db "Mock context" in
```
Comment (gaborigloi) [inline ocaml/xapi/test_vm_check_operation_error.ml]: I'm not sure whether I should use `Test_common.make_test_database` instead.

**Comment by gaborigloi** [comment]
I'm also planning to submit a pull requests that changes the API generation to also generate lists such as `all_vm_operations` for each enum - I don't know any method with which I could enforce the completeness of these lists using the type system. But if you know of a way to do that, please tell me about it!

File: ocaml/xapi/xapi_vm_lifecycle.ml [hunk 2]
```diff
@@ -513,14 +563,7 @@ let update_allowed_operations ~__context ~self =
     | None -> op :: accu
     | _    -> accu
   in
-  let allowed =
-    List.fold_left check []
-      [`snapshot; `copy; `clone; `revert; `checkpoint; `snapshot_with_quiesce;
-       `start; `start_on; `pause; `unpause; `clean_shutdown; `clean_reboot;
-       `hard_shutdown; `hard_reboot; `suspend; `resume; `resume_on; `export; `destroy;
-       `provision; `changing_VCPUs_live; `pool_migrate; `migrate_send; `make_into_template; `changing_static_range;
-       `changing_shadow_memory; `changing_dynamic_range]
```
Comment (thomassa) [inline ocaml/xapi/xapi_vm_lifecycle.ml]: This is deliberately a subset of VM operations.
We do not want to run the checks for all the operations: only the ones for which it is cheap to check whether they are currently allowed.

File: ocaml/xapi/test_vm_check_operation_error.ml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  let vm_ref = Test_common.make_vm ~__context () in
+  f __context vm_ref
+
+(* CA-255128 Xapi_vm_lifecycle.check_operation_error should not throw an
+   exception when one of the nullable fields in the VM record are null. *)
```
Comment (thomassa) [inline ocaml/xapi/test_vm_check_operation_error.ml]: This would be better as `...when the VM has a VBD with a null VDI. *)` rather than the current vague description.
(Also, "when one ... are null" is wrong grammatically.)

Comment (gaborigloi) [inline ocaml/xapi/test_vm_check_operation_error.ml]: What I wanted to say is that the VM class has some fields that can validly be null in practice, i.e. it is not an error if they are null. And when one of these fields are null, the `check_operation_error` function should not fail.  When I create the test VM, the `Test_common.make_vm` function already sets many of the VM's fields to null, so I'm also testing other nullable fields here, not just the VDI. But I like your version, it's more concise, but I think we should still say that this test could fail due to other fields being null as well.

File: ocaml/xapi/xapi_vm_lifecycle.ml (hunk 2 above)
Comment (gaborigloi) [inline ocaml/xapi/xapi_vm_lifecycle.ml]: Hmm, I can't see any expensive check in `Xapi_vm_lifecycle.check_operation_error` that is activated by passing to it a specific operation. I talked to this about @jonludlam and he said that we can just extend it to the full set, to achieve a more correct behaviour, but it might cause bugs. But I'll remove this change if it's a deliberate choice.

File: ocaml/xapi/test_vm_check_operation_error.ml (hunk 3 above)
Comment (thomassa) [inline ocaml/xapi/test_vm_check_operation_error.ml]: Ah, I see. In that case, I suggest
"...should not throw an exception even when all the VM record's nullable fields are null and the VM has a VBD with a null VDI."

Comment (gaborigloi) [inline ocaml/xapi/test_vm_check_operation_error.ml]: Thanks, updated.

File: ocaml/xapi/xapi_vm_lifecycle.ml (hunk 2 above)
Comment (thomassa) [inline ocaml/xapi/xapi_vm_lifecycle.ml]: Having looked at that code, I think I was probably mis-remembering: thinking of something else where we don't update the expensive fields.
